From bb1798120ca44a5632deee5f8a940878b5b97bc9 Mon Sep 17 00:00:00 2001
From: tanghui <tanghui20@huawei.com>
Date: Fri, 22 Apr 2022 11:15:22 +0800
Subject: [PATCH 0792/2065] [Huawei] sched: Adjust cpu allowed in load balance
 dynamicly

Offering: HULK
hulk inclusion
category: feature
bugzilla: 186575

--------------------------------

Not allow task to migrate out of cpu preferred.

Signed-off-by: tanghui <tanghui20@huawei.com>
---
 kernel/sched/fair.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index a729774b54fd..6552766ef795 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -8074,7 +8074,12 @@ int can_migrate_task(struct task_struct *p, struct lb_env *env)
 	if (kthread_is_per_cpu(p))
 		return 0;
 
+#ifdef CONFIG_QOS_SCHED_DYNAMIC_AFFINITY
+	set_task_select_cpus(p, NULL, 0);
+	if (!cpumask_test_cpu(env->dst_cpu, p->select_cpus)) {
+#else
 	if (!cpumask_test_cpu(env->dst_cpu, p->cpus_ptr)) {
+#endif
 		int cpu;
 
 		schedstat_inc(p->se.statistics.nr_failed_migrations_affine);
@@ -8094,7 +8099,11 @@ int can_migrate_task(struct task_struct *p, struct lb_env *env)
 
 		/* Prevent to re-select dst_cpu via env's CPUs: */
 		for_each_cpu_and(cpu, env->dst_grpmask, env->cpus) {
+#ifdef CONFIG_QOS_SCHED_DYNAMIC_AFFINITY
+			if (cpumask_test_cpu(cpu, p->select_cpus)) {
+#else
 			if (cpumask_test_cpu(cpu, p->cpus_ptr)) {
+#endif
 				env->flags |= LBF_DST_PINNED;
 				env->new_dst_cpu = cpu;
 				break;
-- 
2.17.1

